auth/env_keychain.go

```diff
@@ -12,15 +12,30 @@ import (
 	"github.com/pkg/errors"
 )
 
-// NewKeychain returns either a EnvKeychain or a authn.DefaultKeychain depending on whether the provided environment variable is set
-func NewKeychain(envVar string) authn.Keychain {
+// NewKeychain returns an EnvKeychain based on the provided environment variable.
+// If the provided environment variable is not set, NewKeychain will build an auth config
+// from authn.DefaultKeychain and set the variable in the environment.
+// If any errors are encountered, NewKeychain will return an anonymous keychain.
+func NewKeychain(envVar string, images ...string) authn.Keychain {
 	_, ok := os.LookupEnv(envVar)
 	if !ok {
-		return authn.DefaultKeychain
+		authConfig, err := BuildEnvVar(authn.DefaultKeychain, images...)
```

Comment:
If `BuildEnvVar` fails to read the docker config, it will return an empty auth config (instead of erroring). This makes it difficult to log at this stage.

Reply:
Note that we already warn when we fail to find the app image.

Reply:
I am okay omitting the warning entirely for now. I only really wanted to warn if something strange has happened (e.g. there is a docker `config.json` but we can't read it).

In addition I agree that we should add a log message for a missing cache image, as you suggested. However, I think it should be an info message rather than a warning because the cache image would normally be non-existent on a first build and we shouldn't produce warnings in a happy path situation.

Reply:
We looked into this a bit more - the line that I linked to was incorrect because we don't have a logger at that point in the code and furthermore, `DecodeLabel` does not error if the image is not found. We extended the cache interface with an `Exists()` method to get this information at the point where we have a logger.